addons/l10n_my_edi_extended/tests/test_new_submissions_flow.py

```diff
@@ -0,0 +1,657 @@
+             patch('odoo.addons.l10n_my_edi.models.account_move.SUBMISSION_MAX_SIZE', 2):
+            self.submission_invoice.action_l10n_my_edi_send_invoice()
+
+        # we have 10 invoices, with a max size of 2 we expect 5 different submissions.
+        self.assertEqual(self.submission_count, 5)
+
+    @freeze_time('2024-07-15 10:00:00')
+    def test_09_new_fetch_status(self):
+        """ After pushing an invoice, we can optionally fetch the status manually if needed. """
+        with patch(CONTACT_PROXY_METHOD, new=self._test_09_mock):
+            self.basic_invoice.action_l10n_my_edi_send_invoice()
+
+            self.assertRecordValues(
+                self.basic_invoice,
+                [{
+                    'l10n_my_edi_state': 'in_progress',
+                    'l10n_my_edi_submission_uid': '123456789',
+                    'l10n_my_edi_external_uuid': '123458974513518',
+                }]
+            )
+
+            # ... some time later, the user does not want to wait for the cron and press the button.
+            self.basic_invoice.action_l10n_my_edi_update_status()
+
+            # The update should be reflected on the move.
+            self.assertEqual(self.basic_invoice.l10n_my_edi_state, 'valid')
+
+    @freeze_time('2024-07-15 10:00:00')
+    def test_10_new_full_rejection_flow_invoice(self):
+        """
+        We issue an invoice to our customer with the wrong address.
+        The customer reject it for that reason.
+        We receive the updated status later on, and cancel the invoice to issue a new one later.
+        """
+        with patch(CONTACT_PROXY_METHOD, new=self._test_10_mock):
+            # Issue the invoice, and get a valid status.
+            self.basic_invoice.action_l10n_my_edi_send_invoice()
+            # Update the status, and receive a rejection request.
+            self.basic_invoice.action_l10n_my_edi_update_status()
+            self.assertEqual(self.basic_invoice.l10n_my_edi_state, 'rejected')
+            # We then cancel the invoice as the reason is valid.
+            action = self.basic_invoice.button_request_cancel()
+            wizard = self.env[action['res_model']].with_context(action['context']).create({
+                'reason': 'Wrong address',
+            })
+            # Cancel the invoice
+            wizard.button_request_update()
+            self.assertEqual(self.basic_invoice.l10n_my_edi_state, 'cancelled')
+
+    @freeze_time('2024-07-15 10:00:00')
+    def test_11_qr_code_generation(self):
+        """ Basic test that ensure that a valid invoice can generate a QR code. """
+        with patch(CONTACT_PROXY_METHOD, new=self._test_11_mock):
+            self.basic_invoice.action_l10n_my_edi_send_invoice()
+
+        qr_data_uri = self.basic_invoice._generate_myinvois_qr_code()
+        self.assertTrue(qr_data_uri)
+
+    # -------------------------------------------------------------------------
+    # Patched methods
```

Comment:
Do we also want external tests to make sure the API doesn't change?

Reply:
The API that we contact from Odoo is our own (IAP service) so I think this is not necessary (hopefully whoever would change API's endpoints on IAP would think about the Odoo side :sweat_smile: )